src/blockchain_utilities/blockchain_blackball.cpp

```diff
@@ -0,0 +1,403 @@
+  if (boost::filesystem::exists(direc))
+  {
+    if (!boost::filesystem::is_directory(direc))
+    {
+      MERROR("LMDB needs a directory path, but a file was passed: " << output_file_path.string());
+      return 1;
+    }
+  }
+  else
+  {
+    if (!boost::filesystem::create_directories(direc))
+    {
+      MERROR("Failed to create directory: " << output_file_path.string());
+      return 1;
+    }
+  }
+
+  LOG_PRINT_L0("Scanning for blackballable outputs...");
+
+  size_t done = 0;
+  std::unordered_map<crypto::key_image, std::vector<uint64_t>> relative_rings;
+  std::map<output_data, std::unordered_set<crypto::key_image>> outputs;
+  std::set<output_data> spent, newly_spent;
+  tools::ringdb ringdb(output_file_path.string());
+
+  for (size_t n = 0; n < inputs.size(); ++n)
+  {
+    LOG_PRINT_L0("Reading blockchain from " << inputs[n]);
+    for_all_transactions(inputs[n], [&](const cryptonote::transaction_prefix &tx)->bool
+    {
+      for (const auto &in: tx.vin)
+      {
+        if (in.type() != typeid(txin_to_key))
+          continue;
+        const auto &txin = boost::get<txin_to_key>(in);
+        if (opt_rct_only && txin.amount != 0)
+          continue;
+
+        const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(txin.key_offsets);
+        if (n == 0)
+          for (uint64_t out: absolute)
+            outputs[output_data(txin.amount, out)].insert(txin.k_image);
+
+        const uint32_t ring_size = txin.key_offsets.size();
+        if (ring_size == 1)
+        {
+          const crypto::public_key pkey = core_storage[n]->get_output_key(txin.amount, txin.key_offsets[0]);
+          MINFO("Blackballing output " << pkey << ", due to being used in a 1-ring");
+          ringdb.blackball(pkey);
+          newly_spent.insert(output_data(txin.amount, txin.key_offsets[0]));
+          spent.insert(output_data(txin.amount, txin.key_offsets[0]));
+        }
+        else if (relative_rings.find(txin.k_image) != relative_rings.end())
+        {
+          MINFO("Key image " << txin.k_image << " already seen: rings " <<
+              boost::join(relative_rings[txin.k_image] | boost::adaptors::transformed([](uint64_t out){return std::to_string(out);}), " ") <<
+              ", " << boost::join(txin.key_offsets | boost::adaptors::transformed([](uint64_t out){return std::to_string(out);}), " "));
+          if (relative_rings[txin.k_image] != txin.key_offsets)
+          {
+            MINFO("Rings are different");
+            const std::vector<uint64_t> r0 = cryptonote::relative_output_offsets_to_absolute(relative_rings[txin.k_image]);
+            const std::vector<uint64_t> r1 = cryptonote::relative_output_offsets_to_absolute(txin.key_offsets);
+            std::vector<uint64_t> common;
+            for (uint64_t out: r0)
+            {
+              if (std::find(r1.begin(), r1.end(), out) != r1.end())
+                common.push_back(out);
+            }
```

Comment:
How about using `std::set_intersection`?

```c++
std::set_intersection(r1.begin(), r1.end(), r2.begin(), r2.end(), std::back_inserter(common));
```

Reply:
Mostly because I did not know that existed. At least I know what the current code does. The "equivalent code" in the docs isn't immediately obviously the same as mine, so I'm keeping mine.